src/pocketmine/event/entity/EntityDamageEvent.php

```diff
@@ -303,4 +312,30 @@ public function useArmors(){
 		}
 		return false;
 	}
+
+	public function createThornsDamage(){
+		if($this->thornsLevel !== []){
+			$this->thornsArmor = array_rand($this->thornsLevel);
+			$ThronsL = $this->thornsLevel[$this->thornsArmor];
+			if(mt_rand(1, 100) < $ThronsL * 15){
+				$this->thornsDamage = mt_rand(1, 4);
+			}
+		}
+	}
+
+	public function getThornsDamage(){
+		return $this->thornsDamage;
+	}
+
+	/**
+	 * @return bool should be used after getThornsDamage()
+	 */
+	public function setThornsArmorUse(){
+		if($this->thornsArmor === unll){
```

Comment:
Right.
1. $this->thornsArmor = unll (when event was created)
2. in createThornsDamage()
$this->thornsArmor = array_rand($this->thornsLevel);
3. If thorns trigger then setThornsArmorUse.
PS  setThornsArmorUse  mean set the thorns enchantment armor lost 3 durability
PS2 thorns in WIKI
Damages attackers
(Level × 15)% chance of inflicting 1 (Half Heart.svg)-4 (Heart.svgHeart.svg) (or Level - 10 if Level > 10) damage on anyone who attacks wearer, whether by melee attack, arrow, or other projectile. In addition to the normal durability reduction for being hit, reduces durability by 3 points when inflicting damage and 1 point otherwise.
If present on multiple pieces of armor, each armor piece in turn will be given its chance to inflict Thorns damage on the attacker, but the extra Thorns durability penalty will be applied to one of those armor pieces chosen randomly.